azure_data_cosmos_engine/src/query/node.rs

```diff
@@ -169,3 +165,81 @@ impl PipelineNode for OffsetPipelineNode {
         rest.run()
     }
 }
+
+pub struct AggregatePipelineNode {
+    aggregators: Vec<Aggregator>,
+    results: Option<VecDeque<Box<serde_json::value::RawValue>>>,
+}
+
+impl AggregatePipelineNode {
+    pub fn from_names(names: Vec<String>) -> crate::Result<Self> {
+        let mut aggregators = Vec::with_capacity(names.len());
+        for name in names {
+            aggregators.push(Aggregator::from_str(&name)?);
+        }
+        Ok(Self {
+            aggregators,
+            results: None,
+        })
+    }
+}
+
+impl PipelineNode for AggregatePipelineNode {
+    fn next_item(&mut self, mut rest: PipelineSlice) -> crate::Result<PipelineNodeResult> {
+        fn drain_result(
+            results: &mut VecDeque<Box<serde_json::value::RawValue>>,
+        ) -> crate::Result<PipelineNodeResult> {
+            if let Some(value) = results.pop_front() {
+                Ok(PipelineNodeResult::result(
+                    QueryResult {
+                        payload: Some(value),
+                        ..Default::default()
+                    },
+                    results.is_empty(),
+                ))
+            } else {
+                // We shouldn't be here if there are no results left.
+                Ok(PipelineNodeResult {
+                    value: None,
+                    terminated: true,
+                })
+            }
+        }
+
+        if let Some(results) = &mut self.results {
+            return drain_result(results);
+        }
+
+        let result = rest.run()?;
+        if let Some(item) = result.value {
+            tracing::debug!(aggregator_count = self.aggregators.len(), "processing item");
+            for clause_item in item.aggregates {
+                for aggregator in &mut self.aggregators {
+                    aggregator.aggregate(&clause_item)?
+                }
+            }
+        }
+
+        if result.terminated {
+            tracing::debug!("aggregation complete, producing final result");
+            let mut results = VecDeque::with_capacity(self.aggregators.len());
+            for aggregator in self.aggregators.drain(..) {
+                let value = aggregator.into_value()?;
+                if let Some(value) = value {
+                    let value = serde_json::value::to_raw_value(&value).map_err(|e| {
+                        ErrorKind::InternalError
+                            .with_message(format!("failed to serialize aggregate result: {}", e))
+                    })?;
+                    results.push_back(value);
```

Comment:
Lots of shadowing of `value` here. I guess it's ok if idiomatic, but can be confusing what `value` actually is in different places.